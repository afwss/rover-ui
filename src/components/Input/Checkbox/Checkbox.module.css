.outline {
  fill: gray;
}

.inner {
  fill: white;
}

.check,
.dash {
  display: none;
}

.checked .outline {
  fill: blue;
}

.checked .inner {
  display: none;
}

.checked .check {
  display: initial;
  fill: white;
}

.indeterminate .dash {
  display: initial;
  fill: white;
}
