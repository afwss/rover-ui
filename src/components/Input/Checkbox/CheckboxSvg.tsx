import React from 'react';

import styles from './Checkbox.module.css';

const svg = ({ title = 'Checkbox', ...passedProps }) => (
  <svg
    width="18px"
    height="18px"
    viewBox="0 0 18 18"
    version="1.1"
    xmlns="http://www.w3.org/2000/svg"
    {...passedProps}
  >
    <title>{title}</title>
    <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
      <polygon
        className={styles.outline}
        fill="#34404B"
        points="0 0 18 0 18 18 0 18"
      />
      <polygon
        className={styles.inner}
        fill="#34404B"
        points="16 16 16 2 2 2 2 16"
      />
      <polygon
        className={styles.check}
        fill="#34404B"
        points="7 14 2 9 3.41 7.59 7 11.17 14.59 3.58 16 5"
      />
      <polygon
        className={styles.dash}
        fill="#34404B"
        points="14 10 4 10 4 8 14 8"
      />
    </g>
  </svg>
);

export default svg;
