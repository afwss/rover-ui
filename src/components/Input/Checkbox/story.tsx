import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean, text } from '@storybook/addon-knobs';

import Checkbox from '.';
import Readme from './README.md';

import { Wrap } from '../../../stories/storybook-helpers';

interface InteractiveInputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  onChange: () => void;
  InputRenderer: React.FC<React.InputHTMLAttributes<HTMLInputElement>>;
}

const InteractiveInput: React.FC<InteractiveInputProps> = ({
  InputRenderer,
  onChange,
  checked: defaultChecked = undefined,
  ...passedProps
}) => {
  const [value, setValue] = useState('');
  const [checked, setChecked] = useState<boolean | undefined>(defaultChecked);

  return (
    <InputRenderer
      {...passedProps}
      checked={checked !== undefined ? checked : undefined}
      onChange={(e) => {
        setValue(e.target.value);
        setChecked(e.target.checked);
        onChange();
      }}
      value={value !== undefined ? value : undefined}
    />
  );
};

storiesOf('Planets/Input/Checkbox', module)
  .addParameters({
    readme: {
      sidebar: Readme,
    },
  })
  .add(
    'Overview',
    () => (
      <Wrap>
        {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
        <label className="text-xl inline-block mb-2">
          {`<Checkbox />`}
          <p className="mt-6">
            <small>
              <em>
                Checkbox is a bare input. You&apos;ll have to provide your own
                label. &ldquo;Control&rdquo; components with integrated labels
                and validation messages are on the roadmap.
              </em>
            </small>
          </p>
          <div>
            <Checkbox
              checked={boolean('checked', false)}
              className={text('className (HTML)', 'w-full')}
              fauxDisabled={boolean('fauxDisabled', false)}
              id="SamplCheckbox"
              onChange={action('onChange (HTML)')}
            />
          </div>
        </label>
      </Wrap>
    ),
    {
      info: {
        inline: true,
        source: true,
      },
    }
  )
  .add(
    'Examples',
    () => (
      <Wrap>
        <h2>Checkbox</h2>
        <p className="mt-6">
          <small>
            <em>
              Checkbox is a bare input - the labels are provided for
              documentation purposes only. &ldquo;Control&rdquo; components with
              integrated labels and validation messages are on the roadmap.
            </em>
          </small>
        </p>
        <p className="mt-6">
          <small>
            This component is identical to {'`<Input type="checkbox" />`'}.
          </small>
        </p>
        <p className="mt-6">
          <small>
            These examples use an `InteractiveInput` wrapper with an
            `InputRenderer` prop. In actual implementations, just use a
            `Checkbox` component, `onChange`, and `checked` props instead.
          </small>
        </p>
        <div>
          {[
            [],
            ['checked'],
            ['disabled'],
            ['fauxDisabled'],
            ['checked', 'disabled'],
          ].map((attributes) => (
            <div className="my-6" key={['atts', ...attributes].join('-')}>
              <label htmlFor={['atts', ...attributes].join('-')}>
                {attributes.join(' and ') || 'default'}{' '}
                <InteractiveInput
                  InputRenderer={Checkbox}
                  id={['atts', ...attributes].join('-')}
                  onChange={action(
                    `onChange ${attributes.join(' and ') || 'default'}`
                  )}
                  {...attributes.reduce(
                    (acc, attr) => ({
                      ...acc,
                      [attr]: true,
                    }),
                    {}
                  )}
                />
              </label>
            </div>
          ))}
        </div>
      </Wrap>
    ),
    {
      info: {
        inline: false,
        source: true,
      },
    }
  );
