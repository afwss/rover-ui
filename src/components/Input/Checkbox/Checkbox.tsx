import React from 'react';

import classNames from 'classnames';

import type { InputProps } from '../Input';

import CheckboxSvg from './CheckboxSvg';
import styles from './Checkbox.module.css';

interface CheckboxWithRefProps extends InputProps {
  fauxDisabled?: boolean;
  forwardedRef?: React.Ref<HTMLInputElement>;
}

const CheckboxWithRef: React.FC<CheckboxWithRefProps> = ({
  checked = false,
  className = '',
  fauxDisabled = false,
  forwardedRef: ref,
  ...passedProps
}) => {
  return (
    <div
      className={classNames(styles.Checkbox, className, {
        [styles.checked]: checked,
        [styles.disabled]: fauxDisabled,
      })}
    >
      <input
        checked={checked}
        className={styles.input}
        ref={ref || undefined}
        tabIndex={fauxDisabled ? -1 : undefined}
        type="checkbox"
        {...passedProps}
      />
      <CheckboxSvg aria-hidden className={styles.svg} />
    </div>
  );
};

const Checkbox = React.forwardRef<HTMLInputElement, InputProps>(
  (props, ref) => <CheckboxWithRef {...props} forwardedRef={ref || undefined} />
);

export default Checkbox;
