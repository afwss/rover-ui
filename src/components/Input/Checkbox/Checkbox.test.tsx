import React, { useEffect, useRef } from 'react';

import { render } from '@testing-library/react';

import Checkbox from './Checkbox';

/* eslint-disable @typescript-eslint/no-explicit-any */
const WithUseRef = ({ Component, focus = false, ...passedProps }) => {
  const ref = useRef<any>(null);

  useEffect(() => {
    if (ref && ref.current) {
      if (focus) {
        ref.current.focus();
      } else {
        ref.current.blur();
      }
    }
  }, [focus]);

  return <Component ref={ref} {...passedProps} />;
};
/* eslint-disable @typescript-eslint/no-explicit-any */

describe('Input', () => {
  it('renders', () => {
    const { baseElement } = render(<Checkbox onChange={() => {}} />);
    expect(baseElement).toMatchSnapshot();
  });

  describe('Supports ref forwarding', () => {
    it('with useRef', () => {
      const { getByTestId, rerender } = render(
        <WithUseRef
          Component={Checkbox}
          data-testid="my-checkbox"
          onChange={() => {}}
        />
      );

      const checkbox = getByTestId('my-checkbox') as HTMLInputElement;

      expect(checkbox).not.toBe(document.activeElement);

      rerender(<WithUseRef Component={Checkbox} focus />);

      expect(checkbox).toBe(document.activeElement);
    });

    it('with React.createRef', () => {
      const ref = React.createRef<any>();

      const { getByTestId } = render(
        <Checkbox ref={ref} data-testid="my-checkbox" onChange={() => {}} />
      );

      const checkbox = getByTestId('my-checkbox') as HTMLInputElement;
      expect(checkbox).not.toBe(document.activeElement);
      checkbox.focus();
      expect(checkbox).toBe(document.activeElement);
    });

    it('with callback refs', () => {
      const handleSetRef = jest.fn(() => {});

      render(<Checkbox ref={handleSetRef} onChange={() => {}} />);

      const {
        mock: { calls },
      } = handleSetRef;

      let firstArgs: any = calls.length ? calls[0] : [];
      firstArgs = firstArgs.length ? firstArgs[0] : null;
      expect(handleSetRef.mock.calls.length).toBe(1);
      expect(firstArgs.toString()).toBe('[object HTMLInputElement]');
    });
  });
});
