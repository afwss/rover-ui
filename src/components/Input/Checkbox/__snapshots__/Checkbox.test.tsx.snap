// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders 1`] = `
<body>
  <div>
    <div
      class="Checkbox"
    >
      <input
        class="input"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="svg"
        height="18px"
        version="1.1"
        viewBox="0 0 18 18"
        width="18px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Checkbox
        </title>
        <g
          fill="none"
          fill-rule="evenodd"
          stroke="none"
          stroke-width="1"
        >
          <polygon
            class="outline"
            fill="#34404B"
            points="0 0 18 0 18 18 0 18"
          />
          <polygon
            class="inner"
            fill="#34404B"
            points="16 16 16 2 2 2 2 16"
          />
          <polygon
            class="check"
            fill="#34404B"
            points="7 14 2 9 3.41 7.59 7 11.17 14.59 3.58 16 5"
          />
          <polygon
            class="dash"
            fill="#34404B"
            points="14 10 4 10 4 8 14 8"
          />
        </g>
      </svg>
    </div>
  </div>
</body>
`;
